From 641f51738dc3f884477e01492fedd5d52603cd0d Mon Sep 17 00:00:00 2001
From: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
Date: Fri, 21 Jun 2013 14:36:01 +0900
Subject: [PATCH 655/715] ARM: shmobile: r8a7790: Add shdma-desc clock support

Add the module clock setting for SYS-DMAC.

Signed-off-by: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
(cherry picked from commit e3149041d09504d13eff6a4498d4642307948c6e)

Conflicts:
	arch/arm/mach-shmobile/clock-r8a7790.c

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 arch/arm/mach-shmobile/clock-r8a7790.c |    7 ++++++-
 arch/arm/mach-shmobile/setup-r8a7790.c |   11 +++++++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-shmobile/clock-r8a7790.c b/arch/arm/mach-shmobile/clock-r8a7790.c
index b64648a..0f6bc2c 100644
--- a/arch/arm/mach-shmobile/clock-r8a7790.c
+++ b/arch/arm/mach-shmobile/clock-r8a7790.c
@@ -190,8 +190,9 @@ enum {
 	MSTP315, MSTP314, MSTP313, MSTP312, MSTP311, MSTP305, MSTP304,
 	MSTP216, MSTP207, MSTP206, MSTP204, MSTP203, MSTP202,
 	MSTP112,
+	MSTP219, MSTP218,
 	MSTP502, MSTP501,
-	MSTP704, MSTP703, 
+	MSTP704, MSTP703,
 	MSTP917, MSTP931, MSTP930, MSTP929, MSTP928, MSTP922,
 	MSTP1031, MSTP1030, MSTP1019, MSTP1018, MSTP1017, MSTP1015, MSTP1014,
 	MSTP1005,
@@ -222,6 +223,8 @@ static struct clk mstp_clks[MSTP_NR] = {
 	[MSTP717] = SH_CLK_MSTP32(&zs_clk, SMSTPCR7, 17, 0), /* HSCIF0 */
 	[MSTP716] = SH_CLK_MSTP32(&zs_clk, SMSTPCR7, 16, 0), /* HSCIF1 */
 	[MSTP112] = SH_CLK_MSTP32(&zg_clk, SMSTPCR1, 12, 0),
+	[MSTP219] = SH_CLK_MSTP32(&hp_clk, SMSTPCR2, 19, 0),
+	[MSTP218] = SH_CLK_MSTP32(&hp_clk, SMSTPCR2, 18, 0),
 	[MSTP502] = SH_CLK_MSTP32(&hp_clk, SMSTPCR5, 2, 0),
 	[MSTP501] = SH_CLK_MSTP32(&hp_clk, SMSTPCR5, 1, 0),
 	[MSTP704] = SH_CLK_MSTP32(&mp_clk, SMSTPCR7, 04, 0),
@@ -281,6 +284,8 @@ static struct clk_lookup lookups[] = {
 
 	/* MSTP */
 	CLKDEV_DEV_ID("pvrsrvkm", &mstp_clks[MSTP112]),
+	CLKDEV_CON_ID("sysdmac_lo", &mstp_clks[MSTP219]),
+	CLKDEV_CON_ID("sysdmac_up", &mstp_clks[MSTP218]),
 	CLKDEV_CON_ID("audmac_lo", &mstp_clks[MSTP502]),
 	CLKDEV_CON_ID("audmac_up", &mstp_clks[MSTP501]),
 	CLKDEV_ICK_ID("lvds.0", "rcar-du-r8a7790", &mstp_clks[MSTP726]),
diff --git a/arch/arm/mach-shmobile/setup-r8a7790.c b/arch/arm/mach-shmobile/setup-r8a7790.c
index 91467ba..1fd9423 100644
--- a/arch/arm/mach-shmobile/setup-r8a7790.c
+++ b/arch/arm/mach-shmobile/setup-r8a7790.c
@@ -1127,6 +1127,16 @@ static bool sysdma_filter(struct platform_device *pdev)
 	return true;
 }
 
+static struct clk *sysdma_clk_get(struct platform_device *pdev)
+{
+	if (pdev->id == SHDMA_DEVID_SYS_LO)
+		return clk_get(NULL, "sysdmac_lo");
+	else if (pdev->id == SHDMA_DEVID_SYS_UP)
+		return clk_get(NULL, "sysdmac_up");
+	else
+		return NULL;
+}
+
 static const struct sh_dmadesc_slave_config r8a7790_sysdma_slaves[] = {
 	{
 		.slave_id	= SHDMA_SLAVE_SDHI0_TX,
@@ -1203,6 +1213,7 @@ static struct sh_dmadesc_pdata sysdma_platform_data = {
 	.dmaor_init	= DMAOR_DME,
 	.chclr_present	= 1,
 	.dma_filter	= sysdma_filter,
+	.clk_get	= sysdma_clk_get,
 };
 
 static struct resource r8a7790_sysdmal_resources[] = {
-- 
1.7.10.4

